### ISSUE TITLE
Add a label function as like yup

### ISSUE THREAD

**Issue by yatendra121** [colinhacks/zod#1767]
I have used YUP which provides a label for changing the value of key.
like address: yup.string().required().max(150).label('Address')

But Zod has not any simple function as like that.

**Comment by StefanTerdell** [comment]
Maybe you could use `.describe()`? It's not in the docs, but there's an open issue for that #1443

**Comment by JacobWeisenburger** [comment]
> https://github.com/jquense/yup#schemalabellabel-string-schema
Overrides the key name which is used in error messages.



https://github.com/jquense/yup#tuple

**Comment by JacobWeisenburger** [comment]
https://github.com/jquense/yup#schemalabellabel-string-schema
> `Schema.label(label: string): Schema`
>
> Overrides the key name which is used in error messages.

https://github.com/jquense/yup#tuple
> ```ts
> import { tuple, string, number, InferType } from 'yup';
>
> let schema = tuple([
>   string().label('name'),
>   number().label('age').positive().integer(),
> ]);
> 
> await schema.validate(['James', 3]); // ['James', 3]
> 
> await schema.validate(['James', -24]); // => ValidationError: age must be a positive number
> 
> InferType<typeof schema> // [string, number] | undefined
>
> ```

---

I'm not sure what `.describe()` does, but I don't think it's the same.

I don't see `foo` in the error
```ts
const result = z.string().min( 1 ).describe( 'foo' ).safeParse( '' )
!result.success && console.log( result.error.issues[ 0 ] )
// {
//     code: 'too_small',
//     minimum: 1,
//     type: 'string',
//     inclusive: true,
//     exact: false,
//     message: 'String must contain at least 1 character(s)',
//     path: []
// }
```

**Comment by StefanTerdell** [comment]
Ah, no, that's my bad, I misunderstood the question. While `.describe()` would be analog to `"description"` in JSON schema I guess this would be more like `"title"` along with the added function of shaping the default error messages. Neat idea.

**Comment by JacobWeisenburger** [comment]
PRs are always welcome. No need to wait for confirmation.

**Comment by RG7279805** [comment]
@JacobWeisenburger i need your opinion before i continue with this PR
as i mentioned should the label be as a method to class or its ok as [RawCreateParams](https://github.com/colinhacks/zod/blob/master/src/types.ts#L99) ?

**Comment by JacobWeisenburger** [comment]
I don't know that I have enough experience to make a decision about this. Perhaps @colinhacks would be a better person to ask.

**Comment by RG7279805** [comment]
@JacobWeisenburger any idea what would be the best way to approach him?

**Comment by JacobWeisenburger** [comment]
Perhaps discord?
https://discord.gg/RcG33DQJdf

**Comment by RG7279805** [comment]
@JacobWeisenburger i have Messaged him on discord but i don't think he is gonna respond Nonetheless i will wait

**Comment by JacobWeisenburger** [comment]
Sorry. I wish I had a better answer for you.

**Comment by colinhacks** [comment]
I don't love this API - I think it makes the process of "computing" error messages even more inscrutable. What does it even mean to call this method on a ZodObject schema? Or a ZodEffects? Much easier to specify a custom error message at the time of declaration.

```ts
z.string({ required_error: "Name must be defined", invalid_type_error: "Name must be a string" })
```

To make that less verbose, write a helper that returns `{ required_error: string; invalid_type_error: string }`.

**Comment by RG7279805** [comment]
@colinhacks, as you asked i, have made a working draft https://github.com/colinhacks/zod/pull/1902 (i messed up linking PR to this issue)
```
z.object({
  first: z.string({ label: "First Name", }).min(3),
})
```
it's only working for `z.object({
  first: z.string({ label: "First Name", }).min(3),
})` at the moment but you will get an idea of how this is gonna + I have also added labeled messages in ["en.ts"](https://github.com/RG7279805/zod/blob/label-in-raw-params/src/locales/en.ts) (all the messages are not good)

In answering your question about DX if we add labeled errors I think it will be great. when I used Zod for the first time this was a single feature I thought must exist to make this library complete

### Without Labeled error
```typescript
z.object({
  first: z.string().min(3),
}).safeParse({ first: "AB" })
// error message
// {
//   "code": "too_small",
//   "minimum": 3,
//   "type": "string",
//   "inclusive": true,
//   "exact": false,
//   "message": "Invalid input",
//   "path": [
//     "first"
//   ]
// }
```
we cant send it to a client or show it to a user directly we have to add customized errors everywhere even though if its a really simple message like "First name is Required" since "Invalid input" cant be sent
```typescript
z.object({
  first: z.string({ invalid_type_error: "first name should be of String", required_error: "First name is required" }).min(3, { message: "first name should be more than 3 char" }),
})
```
just to add really simple messages it becomes mandatory to include a "message" for every check

### With Labeled Errors
if the label is passed all those messages can be removed and that error can be passed directly to the client/user
even if they want to customize the message further they can do so
```typescript

z.object({
  first: z.string({ label: "First Name" }).min(3),
}).safeParse({ first: "AB" })
// {
//   "code": "too_small",
//   "minimum": 3,
//   "type": "string",
//   "inclusive": true,
//   "exact": false,
//   "message": "First Name must be at least 3 character(s)",
//   "label": "First Name",
//   "path": [
//     "first"
//   ]
// }


```

**Comment by hgenru** [comment]
> Which can be made less verbose with a helper function that returns `{ required_error: string; invalid_type_error: string }`. It's already pretty clean and straightforward IMO so I'm hesistant to introduce additional complexity here unless there's a dramatic DX improvement to be had.

I agree that it is better to point out custom errors wherever possible. But this is very difficult in an internationalized application. In that vein, it is convenient to have a global error translation and the ability to specify key labels.

**Comment by WilliamABradley** [comment]
@colinhacks @RG7279805 any updates on this proposal?

Currently I'm using a rather brittle workaround for the lack of this feature:

```
/**
 * Stop-gap until https://github.com/colinhacks/zod/pull/1902 is merged.
 * @param schema Schema to label
 * @param label Label to apply
 * @returns Schema with label
 */
export function LabelModel<TSchema extends z.ZodTypeAny>(
  label: string,
  schema: TSchema
) {
  return Object.assign(schema, { _label: label }) as TSchema;
}
```

This workaround has the limited scope of determining what model is failing to parse in a distributed system, however, information is lost if you chain additional modifiers onto the schema, so ideally, there was a better way to do it.

**Comment by RG7279805** [comment]
@WilliamABradley i have made a draft PR but haven't got any response back from colinhack

**Comment by mleister97** [comment]
Any updates on this one?

**Comment by darklight9811** [comment]
Really wish we could have this feature, or a method to insert metadata directly into the schema

**Comment by florian2peaches** [comment]
Internationalization is the main concern:
<img width="376" height="31" alt="Image" src="https://github.com/user-attachments/assets/a4c5742c-00d2-4772-bcef-32e6c6869f19" />

**Comment by colinhacks** [comment]
You can use `.describe()` or `.meta()` in Zod 4.
